### PR TITLE
Remove catch_unwind wrapper around the rust query planner

### PR DESCRIPTION
**Pull request by lrlna**
This PR removes catch_unwind around the invocation of the rust query planner. As part of this change, I have changed all applicable `panic!()`s in `apollo-federation`, as well as remove adding a backtrace capture to `SingleFederationErrors`.

#### `panic()!` at the `apollo-federation`

There were several panic!() marcos still in use in `apollo-federation`. Instead of panicking, these functions now return the expected FederationError.

If you're searching in the code, you'll still see a few panics. They are in tests, in macros or function annotated with `#[cfg(test)]`, or in [unused] composition part of the codebase. More specifically:
* in internally used [CLI](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/cli/src/main.rs#L214)
* in tests:
  * [operation tests](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/src/operation/tests/mod.rs)
  * [build query graph tests](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/src/query_graph/build_query_graph.rs#L2448)
  * [api schema tests](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/tests/api_schema.rs)
  * [build query plan support tests](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/tests/query_plan/build_query_plan_support.rs)
  * [context build query plan test](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/tests/query_plan/build_query_plan_tests/context.rs)
  * [fetch operation name tests](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/tests/query_plan/build_query_plan_tests/fetch_operation_names.rs)
* test helpers annotated with `#[cfg(test)]`:
  * [connectors' json selection test helper](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/src/sources/connect/json_selection/apply_to.rs#L116)
  * [connectors selection! macro](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/src/sources/connect/json_selection/helpers.rs#L10)
* [argument merge strategies](https://github.com/apollographql/router/blob/85f99f19e2a2907bd5597687773d68d63644a0c6/apollo-federation/src/schema/argument_composition_strategies.rs) used solely in composition

#### Removing backtrace capture
We were adding backtrace captures to `SingleFederationErrors` in order to help with debugging unexpected errors while were still developing the rust query planner. Capturing backtraces is quite often [a slow and memory intensive process](https://doc.rust-lang.org/stable/std/backtrace/struct.Backtrace.html#method.capture), so we want to remove this additional functionality for rust query planner's GA. Since we now have a lot of confidence in this implementation and have not faced a panic across millions of operations that were tested and compared, this is a safe thing to remove. 

**Note:** The compliance CI is failing at the moment, and will be fixed after #6395 is merged 

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] ~Documentation~ Changeset [^2] completed
- [ ] ~Performance impact assessed and acceptable~
- Tests added and passing[^3]
    - [ ] ~Unit Tests~
    - [ ] ~Integration Tests~
    - [ ] ~Manual Tests~

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by lrlna** [comment]
@TylerBloom I'd like your review on this, since you're working on tidying up the federation errors and I want to make sure this aligns with that you're doing.

@SimonSapin I'd also like your review on this, since you initially set up the catch_unwind and backtrace capture.

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs [hunk 1]
```diff
@@ -1718,12 +1719,12 @@ impl FetchDependencyGraph {
...
             } else {
                 let Some(child_defer_ref) = &child.defer_ref else {
-                    panic!(
+                    return Err(internal_error!(
                         "{} has defer_ref `{}`, so its child {} cannot have a top-level defer_ref.",
                         node.display(node_index),
                         DisplayOption(node.defer_ref.as_ref()),
                         child.display(child_index),
-                    );
+                    ));
```
Comment (TylerBloom) [inline apollo-federation/src/query_plan/fetch_dependency_graph.rs:1727]: Nit: This is the `bail!` macro:
```suggestion
                    bail!(
                        "{} has defer_ref `{}`, so its child {} cannot have a top-level defer_ref.",
                        node.display(node_index),
                        DisplayOption(node.defer_ref.as_ref()),
                        child.display(child_index),
                    );
```

File: apollo-federation/src/query_plan/query_planner.rs [hunk 2]
```diff
@@ -443,10 +444,10 @@ impl QueryPlanner {
             .root_kinds_to_nodes()?
             .get(&normalized_operation.root_kind)
         else {
-            panic!(
+            return Err(internal_error!(
                 "Shouldn't have a {0} operation if the subgraphs don't have a {0} root",
                 normalized_operation.root_kind
-            );
+            ));
```
Comment (TylerBloom) [inline apollo-federation/src/query_plan/query_planner.rs:450]:
```suggestion
            bail!(
                "Shouldn't have a {0} operation if the subgraphs don't have a {0} root",
                normalized_operation.root_kind
            );
```

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs [hunk 3]
```diff
@@ -3402,7 +3402,9 @@ impl DeferTracking {
 
         if let Some(parent_ref) = &defer_context.current_defer_ref {
             let Some(parent_info) = self.deferred.get_mut(parent_ref) else {
-                panic!("Cannot find info for parent {parent_ref} or {label}");
+                return Err(internal_error!(
+                    "Cannot find info for parent {parent_ref} or {label}"
+                ));
```
Comment (TylerBloom) [inline apollo-federation/src/query_plan/fetch_dependency_graph.rs:3407]:
```suggestion
                bail!("Cannot find info for parent {parent_ref} or {label}");
```

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs [hunk 4]
```diff
@@ -2676,10 +2677,9 @@ impl FetchDependencyGraphNode {
         let operation =
             operation_compression.compress(&self.subgraph_name, subgraph_schema, operation)?;
         let operation_document = operation.try_into().map_err(|err| match err {
-            FederationError::SingleFederationError {
-                inner: SingleFederationError::InvalidGraphQL { diagnostics },
-                ..
-            } => FederationError::internal(format!(
+            FederationError::SingleFederationError(SingleFederationError::InvalidGraphQL {
+                diagnostics,
+            }) => FederationError::internal(format!(
```
Comment (TylerBloom) [inline apollo-federation/src/query_plan/fetch_dependency_graph.rs:2682]:
```suggestion
            }) => internal_error!(
```

**Review by TylerBloom — APPROVED** [review]
Federation side LGTM. Added a few suggestion to use `bail!` and `internal_error!` for consistency.

**Comment by TylerBloom** [comment]
> @TylerBloom I'd like your review on this, since you're working on tidying up the federation errors and I want to make sure this aligns with that you're doing.

Ya, this is inline with what I've been working on.

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs (hunk 1 above)
Comment (TylerBloom) [reply apollo-federation/src/query_plan/fetch_dependency_graph.rs:1727]: The original `internal_error!` actually did that! Renee and I agreed that we should have both a `return Err(FederationError::internal(format!(..)))` and just an internal federation error short hand.

Comment (goto-bus-stop) [reply apollo-federation/src/query_plan/fetch_dependency_graph.rs:1727]: I'd like to make `bail!()` panic in debug builds so we get stack traces. It should only be used for internal errors, which are errors that should never happen in normal operation. `internal_error!()` is just for places where we can't use `ensure!()` or `bail!()`

File: apollo-federation/src/query_plan/fetch_dependency_graph.rs (hunk 4 above)
Comment (lrlna) [reply apollo-federation/src/query_plan/fetch_dependency_graph.rs]: I was not able to commit your suggestions via github, not sure why, so it all had to be in a manual commit 7d629059cad60c4f8ccf514607e08fa059a7a4da

**Review by SimonSapin — APPROVED** [review]
Looks good with one minor change

File: apollo-router/src/executable.rs [hunk 5]
```diff
@@ -751,20 +750,10 @@ fn setup_panic_handler() {
         } else {
             tracing::error!("{}", e)
         }
-        if !USING_CATCH_UNWIND.get() {
-            // Once we've panic'ed the behaviour of the router is non-deterministic
-            // We've logged out the panic details. Terminate with an error code
```
[the remark this answers is not in the document]
Comment (lrlna) [reply apollo-router/src/executable.rs:756]: ooops, thank you for catching!

**Pull request merged**